src/drivers/esp32/src/esp_slip.c

```diff
@@ -0,0 +1,315 @@
+/**
+ * ,---------,       ____  _ __
+ * |  ,-^-,  |      / __ )(_) /_______________ _____  ___
+ * | (  O  ) |     / __  / / __/ ___/ ___/ __ `/_  / / _ \
+ * | / ,--´  |    / /_/ / / /_/ /__/ /  / /_/ / / /_/  __/
+ *    +------`   /_____/_/\__/\___/_/   \__,_/ /___/\___/
+ *
+ * Crazyflie control firmware
+ *
+ * Copyright (C) 2021 Bitcraze AB
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, in version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @file esp_slip.c
+ * Protocol for assembling, sending, receiving and decoding SLIP packets to/from the ESP32 ROM bootloader
+ *  
+ */
+
+#include <string.h>
+
+#include "esp_slip.h"
+#include "uart2.h"
+
+#define ESP_OVERHEAD_LEN 8
+
+typedef enum
+{
+  receiveStart,
+  receiveDirection,
+  receiveCommand,
+  receiveSize,
+  receiveValue,
+  receiveData,
+  receiveEnd,
+  error,
+} ESPblReceiveState;
+static ESPblReceiveState espblReceiveState = receiveStart;
+
+static uint8_t slipSize[2];
+static uint8_t slipValue[4];
+static uint8_t slipDataIndex = 0;
+static uint8_t slipSizeIndex = 0;
+static uint8_t slipValueIndex = 0;
+static bool inEscapeSequence = false;
+
+static uint32_t sendSize;
+
+static uint8_t generateChecksum(uint8_t *sendBuffer, esp_slip_send_packet *senderPacket)
+{
+  uint8_t checksum = 0xEF; // seed
+  for (int i = 0; i < senderPacket->dataSize - 16; i++)
+  {
+    checksum ^= sendBuffer[9 + 16 + i];
+  }
+  return checksum;
+}
+
+static void sendSlipPacket(uint32_t size, uint8_t *data, coms_sendbuffer_t sendBufferFn)
+{
+  uint32_t i;
+  static uint8_t dmaSendBuffer[UART2_DMA_BUFFER_SIZE];
```

Comment:
It would be nice to get rid of references to DMA and UART2. The design of using fcn pointers for communication should remove as much as possible of the knowledge of the underlying communication layer. Ideally this file should not have a dependency to DMA or UART2 (the #include earlier in the file). The buffer size must obviously be set in one way or the other but it maybe it would be possible to do it without the dependency?

Reply:
I will pass the buffer size from level up, considering that will have a dependency to UART2 to get the read/write functions. I'll remove references to UART and DMA from variable names

Reply:
The above fix doesn't allow for a static dmaSendBuffer, so after discussion we decided to define a buffer size here and compare it to the UART2 DMA buffer size in the higher level.